### PR TITLE
Upgrading ocaml to 5.1.1, promoting annobin and updating llvm

### PR DESCRIPTION
**Pull request by mbykhovtsev-ms**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

As part of 3.0 package upgrade I have upgraded ocaml to version `5.1.1` and upgraded and promoted the depenedcies it has such as `annobin` and the dependency of `annobin` which is `llvm`.

###### Change Log  

- Upgraded ocaml to version `5.1.1` and imported the spec from Fedora. Removed old patches that are not needed
- Promoted annobin from extended to core and upgraded to `12.49`
- Added LLVmgold.so files to be produced by `llvm` package

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Pipeline build id: local and https://dev.azure.com/mariner-org/mariner/_build/results?buildId=554185&view=results

### CODE REVIEW

**Comment by nisamson** [comment]
Is there a reason all three of these are in the same PR? This is a pretty meaty change if they're not all interrelated.

**Comment by mbykhovtsev-ms** [comment]
> Is there a reason all three of these are in the same PR? This is a pretty meaty change if they're not all interrelated.

Oh yes. The new version of `ocaml` depends on `annocheck` subpackage built by `annobin` package, so I had to first promote the `annobin` package to core and upgrade it to the newest version as `ocmal` needs newest. Then the newer version of `annobin` requires some files from `llvm` package which our package is missing, so then I had to add the change to make `llvm` produce those files as well.

So essentially it was a dependency chain of Ocaml -> annobin -> llvm

File: SPECS/ocaml/ocaml.spec [hunk 1]
```diff
@@ -226,38 +411,68 @@ make -j1 all
...
+
+%files rpm-macros
+%{_rpmmacrodir}/macros.ocaml-rpm
+%{_rpmconfigdir}/redhat/ocaml_files.py
+
 
 %changelog
+* Fri Mar 08 2024 Mykhailo Bykhovtsev <mbykhovtsev@microsoft.com> - 5.1.1-1
+- Upgraded ocaml to 5.1.1
+- Importing and adopting spec changes from Fedora.
```
Comment (rikenm1) [inline SPECS/ocaml/ocaml.spec:473]: can you also add (License: MIT) after Fedora

File: SPECS/ocaml/ocaml.spec [hunk 2]
```diff
@@ -1,45 +1,130 @@
...
+# not match the real definitions, leading to lots of -Wlto-type-mismatch
+# warnings.  These change the output of the tests, leading to many failed tests.
+%global _lto_cflags %{nil}
+
+# OCaml has a bytecode backend that works on anything with a C
+# compiler, and a native code backend available on a subset of
+# architectures.  A further subset of architectures support native
+# dynamic linking.
+
+%ifarch %{ocaml_native_compiler}
```
Comment (rikenm1) [inline SPECS/ocaml/ocaml.spec:15]: where is %{ocaml_native_compiler} coming from ?

File: SPECS/annobin/annobin.spec [hunk 3]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
 # availablilty at build time, but this might not match the run time situation.
+# FIXME: Add a --without debuginfod option to forcefully disable the configure
+# time check for debuginfod support.
 %bcond_with debuginfod
 
+# Use "--without clangplugin" to disable the building of the annobin plugin for Clang.
+%bcond_without clangplugin
+
+# Use "--without gccplugin" to disable the building of the annobin plugin for GCC.
+%bcond_without gccplugin
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:32]: gccplugin flag will always be true as we do not pass bcond_without flag for anything.  Do we want to use these bcond_without then?

File: SPECS/annobin/annobin.spec [hunk 4]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
+
+# Use "--without llvmplugin" to disable the building of the annobin plugin for LLVM.
+%bcond_without llvmplugin
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:35]: same here.

File: SPECS/annobin/annobin.spec [hunk 5]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
+Summary:       A tool for checking the security hardening status of binaries
+
+BuildRequires: gcc elfutils elfutils-devel elfutils-libelf-devel rpm-devel make
+
+%if %{with debuginfod}
+BuildRequires: elfutils-debuginfod-client-devel
+%endif
+
+Requires: %{name}-docs = %{version}-%{release}
+Requires: cpio rpm
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:172]: nit. multi line sorted Requires

File: SPECS/annobin/annobin.spec [hunk 6]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
+
+%description annocheck
+Installs the annocheck program which uses the notes generated by annobin to
+check that the specified files were compiled with the correct security
+hardening options.
+
+%package libannocheck
+Summary: A library for checking the security hardening status of binaries
+
+BuildRequires: gcc elfutils elfutils-devel elfutils-libelf-devel rpm-devel make
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:182]: same here. multi-line sorted.

File: SPECS/annobin/annobin.spec [hunk 7]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
+
+%if %{with debuginfod}
+BuildRequires: elfutils-debuginfod-client-devel
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:185]: here as well.

File: SPECS/ocaml/ocaml.spec [hunk 8]
```diff
@@ -99,123 +199,208 @@ compilers, useful for the development of some OCaml applications.
...
 
-find %{buildroot} -type f -name "*.la" -delete -print
+mkdir -p $RPM_BUILD_ROOT%{_rpmmacrodir}
+install -m 0644 %{SOURCE1} $RPM_BUILD_ROOT%{_rpmmacrodir}/macros.ocaml-rpm
 
-# Remove .cmt and .cmti files, for now.  We could package them later.
-# See also: http://www.ocamlpro.com/blog/2012/08/20/ocamlpro-and-4.00.0.html
-find %{buildroot} \( -name '*.cmt' -o -name '*.cmti' \) -a -delete -print
+mkdir -p $RPM_BUILD_ROOT%{_rpmconfigdir}/redhat
+install -m 0644 %{SOURCE2} $RPM_BUILD_ROOT%{_rpmconfigdir}/redhat
```
Comment (rikenm1) [inline SPECS/ocaml/ocaml.spec:323]: is it required to have redhat here?

File: SPECS/annobin/annobin.spec [hunk 9]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
+%global annobin_source_dir %{_usrsrc}/annobin
+
+# Insert patches here, if needed.  Eg:
+# Patch01: annobin-plugin-default-string-notes.patch
 
 #---------------------------------------------------------------------------------
 
+# Make sure that the necessary sub-packages are built.
+
+%if %{with gccplugin}
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:82]: these if statements will always be true as I said above. so probably can just be removed.

File: SPECS/annobin/annobin.spec [hunk 10]
```diff
@@ -34,29 +21,189 @@ URL:     https://fedoraproject.org/wiki/Toolchain/Watermark
...
+Requires: %{name}-plugin-gcc
+%endif
+
+%if %{with llvmplugin}
```
Comment (rikenm1) [inline SPECS/annobin/annobin.spec:86]: same

**Review by rikenm1 — COMMENTED** [review]
Just few questions and some nit

File: SPECS/ocaml/ocaml.spec (hunk 2 above)
Comment (mbykhovtsev-ms) [reply SPECS/ocaml/ocaml.spec:15]: It seems lots of ocaml based packages use it, we actually do have that macro defined in `macros.ocaml-srpm` which is:
```
# Architectures that support the OCaml native code compiler.
%ocaml_native_compiler  aarch64 %{arm} %{ix86} ppc ppc64 ppc64le riscv64 s390x sparc sparcv9 x86_64
```

**Review by rikenm1 — APPROVED** [review]
LGTM

**Pull request merged**